test/base/src/androidTest/kotlin/io/realm/test/shared/RealmInstantTests.kt

```diff
@@ -0,0 +1,143 @@
+            assertEquals(Long.MIN_VALUE, min.epochSeconds)
+            assertEquals(-999_999_999, min.nanosecondsOfSecond)
+            assertEquals(RealmInstant.MIN, min)
+        }
+
+        roundTrip(RealmInstant.fromEpochSeconds(Long.MAX_VALUE, 999_999_999)) { max ->
+            assertEquals(Long.MAX_VALUE, max.epochSeconds)
+            assertEquals(999_999_999, max.nanosecondsOfSecond)
+            assertEquals(RealmInstant.MAX, max)
+        }
+
+        roundTrip(RealmInstant.fromEpochSeconds(Long.MAX_VALUE, Int.MAX_VALUE)) { maxOverflow ->
+            assertEquals(RealmInstant.MAX, maxOverflow)
+        }
+
+        roundTrip(RealmInstant.fromEpochSeconds(Long.MAX_VALUE, 1_000_000_000)) { minOverflow ->
+            assertEquals(RealmInstant.MAX, minOverflow)
+        }
+
+        roundTrip(RealmInstant.fromEpochSeconds(Long.MIN_VALUE, Int.MIN_VALUE)) { maxUnderflow ->
+            assertEquals(RealmInstant.MIN, maxUnderflow)
+        }
+
+        roundTrip(RealmInstant.fromEpochSeconds(Long.MIN_VALUE, -1_000_000_000)) { minUnderflow ->
+            assertEquals(RealmInstant.MIN, minUnderflow)
+        }
+
+        roundTrip(RealmInstant.fromEpochSeconds(0, 0)) { zero ->
+            assertEquals(0, zero.epochSeconds)
+            assertEquals(0, zero.nanosecondsOfSecond)
+        }
+
+        roundTrip(RealmInstant.fromEpochSeconds(0, 1)) { zeroPlusOne ->
+            assertEquals(0, zeroPlusOne.epochSeconds)
+            assertEquals(1, zeroPlusOne.nanosecondsOfSecond)
+        }
+
+        roundTrip(RealmInstant.fromEpochSeconds(0, -1)) { zeroMinusOne ->
+            assertEquals(0, zeroMinusOne.epochSeconds)
+            assertEquals(-1, zeroMinusOne.nanosecondsOfSecond)
+        }
+
+        roundTrip(RealmInstant.fromEpochSeconds(-1, 2_000_000_200)) { crossingZeroFromNegative ->
+            assertEquals(RealmInstant.fromEpochSeconds(1, 200), crossingZeroFromNegative)
+        }
+
+        roundTrip(RealmInstant.fromEpochSeconds(1, -2_000_000_200)) { crossingZeroFromPositive ->
+            assertEquals(RealmInstant.fromEpochSeconds(-1, -200), crossingZeroFromPositive)
+        }
+    }
+
+    // Store value and retrieve it again
+    private fun roundTrip(timestamp: RealmInstant, function: (RealmInstant) -> Unit) {
+
+        // Test unmanaged objects
+        function(timestamp)
+
+        // Test managed objects
+        realm.writeBlocking {
+            val sample = copyToRealm(
+                Sample().apply {
+                    timestampField = timestamp
+                }
+            )
+            val managedTimestamp = objects(Sample::class).first().timestampField
+            function(managedTimestamp)
+            cancelWrite() // So we can use .first()
+        }
+    }
+
+    @Test
+    fun equals() {
+        assertTrue(RealmInstant.fromEpochSeconds(42, 42) == (RealmInstant.fromEpochSeconds(42, 42)))
+        assertFalse(RealmInstant.fromEpochSeconds(0, 0) == (RealmInstant.fromEpochSeconds(42, 42)))
```

Comment:
Maybe also test the two remaining cases when equal on one axis. Few lines of code but could lead to a major mess. Also for `hashCode`.